### PR TITLE
[User Experience] Add audio resample function `paddle.audio.functional.resample`

### PR DESCRIPTION
**Pull request by PlumBlossomMaid**
### PR Category
User Experience

### PR Types
New features

### Description
[used AI Studio] 本PR为paddle.audio.functional模块新增音频重采样函数resample，主要功能如下：

功能概述
· 实现基于sinc插值的高质量音频重采样，支持Hann窗和Kaiser窗
· 支持任意整数采样率之间的转换（原始采样率 -> 目标采样率）
· 支持CPU/GPU硬件加速，提升音频处理效率

API设计
```python
paddle.audio.functional.resample(
    waveform: Tensor,
    orig_freq: int,
    new_freq: int,
    lowpass_filter_width: int = 6,
    rolloff: float = 0.99,
    resampling_method: str = "sinc_interp_hann",
    beta: Optional[float] = None,
    name: Optional[str] = None
) -> Tensor
```

实现方案
· 使用现有API组合实现，不新增底层OP
· 核心依赖：conv1d、三角函数、i0函数等
· 实现步骤：采样率约分 -> 生成sinc插值核 -> 边缘填充 -> 卷积插值 -> 长度裁剪

相关文档
· RFC文档: https://github.com/PaddlePaddle/community/pull/1212
· 修改文件: python/paddle/audio/functional/functional.py

### 是否引起精度变化
否

### CODE REVIEW

File: test/legacy_test/test_audio_functions.py [hunk 1]
```diff
@@ -59,6 +61,99 @@ def get_wav_data(dtype: str, num_channels: int, num_frames: int):
...
+            tensor = tensor.T
+
+        return self.convert_tensor_encoding(tensor, dtype)
+
+    def _test_resample_waveform_accuracy(
+        self, up_scale_factor=None, down_scale_factor=None, resampling_method="sinc_interp_hann", atol=1e-1, rtol=1e-4
+    ):
+        # resample the signal and compare it to the ground truth
+        n_to_trim = 20
+        sample_rate = 1000
```
Comment (zxcd) [inline test/legacy_test/test_audio_functions.py]: suggest using a real voice, such as "https://paddlespeech.cdn.bcebos.com/PaddleAudio/zh.wav"

File: test/legacy_test/test_audio_functions.py [hunk 2]
```diff
@@ -59,6 +61,99 @@ def get_wav_data(dtype: str, num_channels: int, num_frames: int):
...
+        tensor /= 2.0
+        tensor *= scale_factor
+        tensor.clip_(-1.0, 1.0)
+        if not channels_first:
+            tensor = tensor.T
+
+        return self.convert_tensor_encoding(tensor, dtype)
+
+    def _test_resample_waveform_accuracy(
+        self, up_scale_factor=None, down_scale_factor=None, resampling_method="sinc_interp_hann", atol=1e-1, rtol=1e-4
```
Comment (zxcd) [inline test/legacy_test/test_audio_functions.py]: This default accuracy is a bit low. It is recommended that the absolute accuracy be compared to torch, and the diff should be better can achieve atol e-6 on "float32".

File: test/legacy_test/test_audio_functions.py [hunk 3]
```diff
@@ -408,6 +503,49 @@ def test_istft(self, n_fft: int, hop_length: int, window_str: str):
...
+
+    @parameterize(["sinc_interp_hann", "sinc_interp_kaiser"], list(range(1, 20)))
+    def test_resample_waveform_downsample_accuracy(self, resampling_method, i):
+        self._test_resample_waveform_accuracy(down_scale_factor=i * 2, resampling_method=resampling_method)
+
+
+    @parameterize(["sinc_interp_hann", "sinc_interp_kaiser"], list(range(1, 20)))
+    def test_resample_waveform_upsample_accuracy(self, resampling_method, i):
+        self._test_resample_waveform_accuracy(up_scale_factor=1.0 + i / 20.0, resampling_method=resampling_method)
+
```
Comment (zxcd) [inline test/legacy_test/test_audio_functions.py]: Need to add some error check test for asserts and errors.

File: python/paddle/audio/functional/functional.py [hunk 4]
```diff
@@ -335,3 +335,288 @@ def create_dct(
...
+    # Indeed:
+    # y[j + new_freq] = sum_i x[i] sinc(pi * orig_freq * ((i / orig_freq - (j + new_freq) / new_freq))
+    #                 = sum_i x[i] sinc(pi * orig_freq * ((i - orig_freq) / orig_freq - j / new_freq))
+    #                 = sum_i x[i + orig_freq] sinc(pi * orig_freq * (i / orig_freq - j / new_freq))
+    # so y[j+new_freq] uses the same filter as y[j], but on a shifted version of x by `orig_freq`.
+    # This will explain the conv1d after, with a stride of orig_freq.
+    width = math.ceil(lowpass_filter_width * orig_freq / base_freq)
+    # If orig_freq is still big after GCD reduction, most filters will be very unbalanced, i.e.,
+    # they will have a lot of almost zero values to the left or to the right...
+    # There is probably a way to evaluate those filters more efficiently, but this is kept for
```
Comment (zxcd) [inline python/paddle/audio/functional/functional.py]: This type of code annotation seems unnecessary, since we don't have a similar plan

Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 该部分已完成

File: test/legacy_test/test_audio_functions.py (hunk 3 above)
Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 所有异常测试已全部添加，保证覆盖所有非法输入情况

File: test/legacy_test/test_audio_functions.py (hunk 2 above)
Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 这份测试算法就是从torchaudio当中借鉴来的，重采样算法使用有限长度的滤波器（FIR近似），在信号的开始和结束位置，由于缺乏足够的相邻采样点，会产生边界伪影，所以会有“掐头去尾”的情况；sinc插值理论上需要无限长的滤波器才能完美重建，但实际上我们使用有限长度的滤波器（lowpass_filter_width=6），这会导致频率响应不是理想的低通滤波器，有一定的过渡带和纹波；相位失真，特别是在高频部分；数值误差累积，特别是在多次重采样时。由于人耳对轻微的相位失真和幅度误差不敏感，大多数音频处理任务（如语音识别）对这些微小误差有鲁棒性且更精确的算法（如更长的滤波器）会显著增加计算开销，这个误差是可以接受的。

File: test/legacy_test/test_audio_functions.py (hunk 1 above)
Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 已经添加了在真实情况下对数据进行处理的测试。

File: test/legacy_test/test_audio_functions.py (hunk 2 above)
Comment (zxcd) [reply test/legacy_test/test_audio_functions.py]: 但这里的单测出现的意义实际上是对于你复现代码的测试，测试的目的是保证你实现的算法和复现的算法精度对齐。从这个逻辑上来看，对齐的精度应该与其他api算子保持一致。如果仍存在无法对齐的部分，可以进行单独分析。

Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 这里面能用torchaudio吗，如果用torchaudio计算结果进行比对，误差能精确到1e-10

Comment (zxcd) [reply test/legacy_test/test_audio_functions.py]: 如果需要加载torchaudio，可以尝试在[PaConvert](https://github.com/PaddlePaddle/PaConvert)的repo中增加单测

Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 那测试精度，精度对齐这一块在这里就先删掉吧。同时去paconvert那边把精度测试补上。这边的话主要就专注于过CI就好了。😊

Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 或者这边这个留着也可以，毕竟可以增加代码的覆盖率。这边的话进行一个粗略的测试，paconvert那边拿真实数据测试对齐到1e-9

**Comment by PlumBlossomMaid** [comment]
任务看板：

- [x] CI覆盖率达标
- [x] 文档中的代码与实际情况对齐
- [x] 代码风格规范化
- [x] 修正PR书写规范
- [x] docstring里面有.. code-block:: python，找人approve
- [ ] paconvert那边添加与torchaudio精度对齐的测试

**Comment by PlumBlossomMaid** [comment]
更新了几个地方：
1. CI覆盖率达标，`Paddle\test\legacy_test\test_audio_functions.py`中的所有代码对`paddle\audio\functional\functional.py`中的`resample`函数覆盖率达100%
2. 文档中的代码居然也会运行，所以这次我对齐了输出的每一个字符
3. 代码风格通过本地执行`pre-commit run --files xxx`进行规范化，不过似乎也影响到了其他函数中的注释
4. 重新写了一下PR的description，尝试通过CI的测试
5. docstring里面有.. code-block:: python，这个需要找谁approve呀
6. paconvert那边以后再补，等前几个没问题了再说

File: test/legacy_test/test_audio_functions.py (hunk 2 above)
Comment (zxcd) [reply test/legacy_test/test_audio_functions.py]: 可以的，这边可以留个做粗略测试，然后写个注释标明paconvert那边的单测在哪

Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_functions.py]: 我要移除这个准确度测试了，Linux和Windows上这个测试都能过，但是Mac的CI居然还能报错，说明不同平台的硬件计算略有差异。在此情况下，和torchaudio对齐才是最正确的选择，这种粗测的办法还是太粗测了。

**Comment by PlumBlossomMaid** [comment]
刚看了一下最新的CI情况，发现遇到了如下问题，希望可以得到回复：
1. 似乎CI里面测试了一些不是我写的函数，然后报错了
2. 有些地方需要找人评审才能通过
3. 我全程使用Windows 10含GPU的笔记本进行开发测试，本地好好的，但是Windows GPU环境下的CI程序直接挂了
4. 我在CI里面用了Reviewer给的音频URL，但是似乎本地是可以运行的，CI上面直接timeout了 何意味

File: python/paddle/audio/functional/functional.py [hunk 5]
```diff
@@ -37,13 +37,14 @@ def hz_to_mel(freq: _TensorOrFloat, htk: bool = False) -> _TensorOrFloat:
         Union[Tensor, float]: Frequency in mels.
 
     Examples:
-        .. code-block:: pycon
+        .. code-block:: python
```
Comment (zxcd) [inline python/paddle/audio/functional/functional.py]: 为什么要改这里？

File: test/legacy_test/test_audio_resample.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+import itertools
+import os
+import unittest
+import urllib
+
+from parameterized import parameterized
+
+import paddle
+import paddle.audio
```
Comment (zxcd) [inline test/legacy_test/test_audio_resample.py]: 建议只引用一次paddle

File: python/paddle/audio/functional/functional.py (hunk 5 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 运行那个代码格式规范化pre-commit的时候脚本自动给改的。一开始我也纳闷，但是paddle这个项目的代码规范就是这么规定的，说明代码原来注释就不规范，我这次提交PR的时候正好给发现并不小心纠正了。

File: test/legacy_test/test_audio_resample.py (hunk 6 above)
Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_resample.py]: 好的

File: python/paddle/audio/functional/functional.py (hunk 5 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 好奇怪的pre-commit规则……我手动都改过来了。

**Comment by PlumBlossomMaid** [comment]
成功申请到了AI Studio当中的框架开发环境，已经使用该环境进行了单卡、多卡自测，函数功能均正常。

**Comment by zxcd** [comment]
TODO: Windows-Inference CI can not pass when run into `paddle.nn.functional.conv1d(paddle.rand([1,8028]),paddle.rand([1,1,28]),stride=2) `
It should be a CI environment issue and cannot be replicated locally. At present, it is recommended to skip it temporarily and fix it in the CI.

**Review by zxcd — APPROVED** [review]
LGTM, approve for function code part.

**Comment by PlumBlossomMaid** [comment]
@zxcd 这个PR啥时候合入呀？我这边自己写的项目依赖了这个PR，现在每次切换环境都要手动给框架添加这个函数

</img>

**Comment by luotao1** [comment]
@SigureMo Please review for API annotation change.

File: python/paddle/audio/functional/functional.py [hunk 7]
```diff
@@ -335,3 +335,290 @@ def create_dct(
...
+    return resampled
+
+
+def resample(
+    waveform: Tensor,
+    orig_freq: int,
+    new_freq: int,
+    lowpass_filter_width: int = 6,
+    rolloff: float = 0.99,
+    resampling_method: str = "sinc_interp_hann",
```
Comment (SigureMo) [inline python/paddle/audio/functional/functional.py]: 有限值集合用 `Literal`，参考 https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/style_guide_and_references/type_annotations_specification_cn.html

File: python/paddle/audio/functional/functional.py [hunk 8]
```diff
@@ -335,3 +335,290 @@ def create_dct(
...
+    beta: float | None = None,
+    name: str | None = None,
```
Comment (SigureMo) [inline python/paddle/audio/functional/functional.py]: 这种新 API 有必要加 `name` 参数么？感觉 name 给兼容性也带来了不少麻烦，而且在如今动态图 + PIR 下应该是完全没用的了 @zhwesky2010

File: python/paddle/audio/functional/functional.py (hunk 7 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 行我这里稍后改改

File: python/paddle/audio/functional/functional.py (hunk 8 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 说实话这里我也犹豫了一下，但是看到众多API是有name参数的，所以还是加上了。
在paddle2.4的时候，这个name参数对我在导出的静态图里面debug还是起到了一定的帮助作用的，方便我对应静态图和动态图的大体位置，所以最后我决定保留。
在paddle3.0的时候我导出过一个简单的ONNX测试了一下，name参数会影响到ONNX模型对应的算子的命名，我感觉如果是想知道ONNX模型算子对应的动态图的代码的话还是有一点点作用的。

File: test/legacy_test/test_audio_resample.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    def test_resample_waveform_identity_shape(self, resampling_method):
+        sr = 16000
+        waveform = self.get_whitenoise(
+            sample_rate=sr,
+            duration=0.5,
+        )
+        resampled = paddle.audio.functional.resample(
+            waveform, sr, sr, resampling_method=resampling_method
+        )
+        assert resampled.shape[-1] == waveform.shape[-1]
```
[the remark this answers is not in the document]
Comment (PlumBlossomMaid) [reply test/legacy_test/test_audio_resample.py:182]: 如果是补充单测的话，因为需要和torchaudio做对齐，所以需要torchaudio库参与运算。
可是单测运行的时候无法导入torchaudio，同时也不能使用torchaudio提前计算好结果，因为不同平台的torchaudio的计算结果甚至都不是一样的。
其实一开始CI里面还有个cos函数的粗测的来着，结果Mac那边的CI炸了，因为Mac的底层运算可能比较特殊吧，导致计算出来的结果与预期值差别较大，而在Linux和Windows平台上均不会出现该问题。
经过与 @zxcd 协商，torchaudio的对齐我将在PaConvert库中添加单测，只是如果PaConvert库的单测想要运行，就必须paddle有这个`resample`函数。这也是我刚才过来催合入的原因之一。

File: python/paddle/audio/functional/functional.py (hunk 7 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 改完了，已经push。

**Review by SigureMo — DISMISSED** [review]
我没问题了，看看 @zhwesky2010 对是否加 name 参数是否有建议

File: tools/windows/run_unittests.sh [hunk 10]
```diff
@@ -315,6 +315,7 @@ disable_win_trt_test="^test_trt_convert_conv2d$|\
 
 # /*==========Fixed Disabled Windows CUDA11.x inference_api_test(PR-CI-Windows-Inference) unittests=============*/
 disable_win_inference_test="^trt_quant_int8_yolov3_r50_test$|\
+^test_audio_resample$|\
```
Comment (zhwesky2010) [inline tools/windows/run_unittests.sh:318]: windows下精度有问题吗

File: python/paddle/audio/functional/functional.py [hunk 11]
```diff
@@ -335,3 +335,294 @@ def create_dct(
...
+    waveform: Tensor,
+    orig_freq: int,
+    new_freq: int,
+    lowpass_filter_width: int = 6,
+    rolloff: float = 0.99,
+    resampling_method: Literal[
+        "sinc_interp_hann", "sinc_interp_kaiser"
+    ] = "sinc_interp_hann",
+    beta: float | None = None,
+    name: str | None = None,
```
Comment (zhwesky2010) [inline python/paddle/audio/functional/functional.py]: 3.0之后的paddle动态图+PIR都不支持name了，目前name仅在老静态图被用上，但这个API是组合的，并未直接使用name，被组合的API很多也不支持name了，所以即使加上也基本不起作用。

因此不用加name了，避免后续还有维护成本。

File: tools/windows/run_unittests.sh (hunk 10 above)
Comment (PlumBlossomMaid) [reply tools/windows/run_unittests.sh:318]: CI机子有问题，等着修

File: python/paddle/audio/functional/functional.py (hunk 11 above)
Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 行我吃完饭把这个删了

Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 我注意到还有部分API有name参数但是实际上并未使用，这也就意味着这种API的name传了跟没传是一样的。在后续的框架维护过程中，是否应该去掉这种无用的name参数？

Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 已去除name参数。

Comment (zhwesky2010) [reply python/paddle/audio/functional/functional.py]: 是应该去掉

Comment (PlumBlossomMaid) [reply python/paddle/audio/functional/functional.py]: 啊哈，看见paddle的API没有name参数了，多少有点不习惯

**Review by zhwesky2010 — APPROVED** [review]
lgtm for public API

**Review by SigureMo — APPROVED** [review]
